### DISCUSSION TITLE
ruling out sections that have already been annotated

### DISCUSSION THREAD

**Discussion by alecristia** [LAAC-LSCP/ChildProject#235] [Q&A]
for the Lyon annotation campaign we are preparing, it would be smartest to remove from consideration sections that have already been annotated as explained [here](https://github.com/LAAC-LSCP/EL1000/issues/22).

I remember we discussed implementing this via an "exclude" csv, that lists onset/offset of sections to skip. Perhaps I'm not reading [the docs](https://childproject.readthedocs.io/en/latest/samplers.html) right, but I don't see something like that -- is it implemented already?

**Comment by lucasgautheron** [comment]
It was not builtin in the package yet;

One possibility would be to perform some sampling, then removing annotated portions from the segments sampled.
Since the segments sampled would be 30 seconds / 1 minute windows in this case, what's going to happen is that some of these windows will be shorter than the target duration after removing previously annotated portions. 

Another (very similar) possibility would be to do the sampling, then calculate the intersection of sampled segments and annotated segments, then discard those of the sampled segments that intersect "too much" (above some threshold) with previously annotated portions.

**Comment by lucasgautheron** [comment]
PR opened @ https://github.com/LAAC-LSCP/ChildProject/pull/236

This PR provides code to remove portions of the samples that match a given set of segments.
It may not be useful in your case, since you are trying to sample fixed-length windows (1-minute or 30-sec windows)

This piece of code will remove windows that have too much overlap with portions of audio previously annotated by humans.
 I didn't even try to run it so corrections are probably required.

```python

# get sampled windows
sampled = pd.read_csv('sampled_segments.csv')
sampled['set'] = 'sampled'
sampled['time_seek'] = 0
sampled['range_onset'] = sampled['segment_onset']
sampled['range_offset'] = sampled['segment_offset']

# load annotations
project = ChildProject('.')
am = AnnotationManager(project)
am.read()

# compute the intersection between sampled windows and annotations from humans
redundant = am.intersection(
    pd.concat([am.annotation, sampled]),
    ['sampled', 'eaf/an1']
)

# calculate the extent of the intersection for each window that has been partially annotated
redundant = redundant[redundant['set'] == 'sampled']
redundant['extent'] = redundant['range_offset']-redundant['range_onset']
redundant = redundant.groupby(['recording_filename', 'segment_onset', 'segment_offset']).agg(
    redundancy_extent = ('extent', 'sum')
)

sampled = sampled.set_index(['recording_filename', 'segment_onset', 'segment_offset'])
sampled = sampled.merge(redundant, how = 'left', left_index = True, right_index = True)
sampled['redundancy_extent'].fillna(0, inplace = True)

# keep windows for which less than five seconds had been previously annotated
sampled = sampled[sampled['redundancy_extent'] < 5000]
```
